sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/ThroughputEstimator.java

```diff
@@ -55,25 +46,18 @@ public ThroughputEstimator() {
    * @param timeOfRecords the committed timestamp of the records
    * @param bytes the total bytes of the records
    */
+  @SuppressWarnings("nullness") // queue is never null, nor the peeked element
   public void update(Timestamp timeOfRecords, long bytes) {
     synchronized (queue) {
-      BigDecimal bytesNum = BigDecimal.valueOf(bytes);
-      if (startTimeOfCurrentWindow.equals(Timestamp.MIN_VALUE)) {
-        bytesInCurrentWindow = bytesNum;
-        startTimeOfCurrentWindow = timeOfRecords;
-        return;
-      }
-
-      if (timeOfRecords.getSeconds() < startTimeOfCurrentWindow.getSeconds() + 1) {
-        bytesInCurrentWindow = bytesInCurrentWindow.add(bytesNum);
+      if (queue.isEmpty() || timeOfRecords.getSeconds() > queue.peek().getLeft().getSeconds()) {
+        queue.add(new ImmutablePair<>(timeOfRecords, BigDecimal.valueOf(bytes)));
       } else {
-        queue.add(new ImmutablePair<>(startTimeOfCurrentWindow, bytesInCurrentWindow));
-        bytesInQueue = bytesInQueue.add(bytesInCurrentWindow);
-
-        bytesInCurrentWindow = bytesNum;
-        startTimeOfCurrentWindow = timeOfRecords;
+        final ImmutablePair<Timestamp, BigDecimal> pair = queue.remove();
```

Comment:
`queue.remove()` will remove the most recent added element, right?

Reply:
No, you are right, I should be using something else (maybe a stack).

Reply:
This should be `queue.removeLast()`.

Reply:
Updated